src/UI/App.hs

```diff
@@ -31,8 +36,34 @@ drawUI s =
         ComposeEditor -> drawComposeEditor s
         Help -> drawHelp s
 
+class EventHandler (m :: Mode) where
+  handle :: Proxy m -> AppState -> Event -> T.EventM Name (T.Next AppState)
+  fallback :: Proxy m -> AppState -> Event -> T.EventM Name (T.Next AppState)
+
+instance EventHandler 'BrowseMail where
+  handle _ s e = handleEvent (view (asConfig . confIndexView . ivKeybindings) s) listEventDefault s e
+
+instance EventHandler 'SearchMail where
+  handle _ s e = handleEventLensed
+
+instance EventHandler 'ManageTags where
+  handle _ s e = mainEvent s e
+
+dispatch :: EventHandler m => Proxy m -> AppState -> T.BrickEvent Name e -> T.EventM Name (T.Next AppState)
+dispatch m s (T.VtyEvent e) = handle m s e
+
 appEvent :: AppState -> T.BrickEvent Name e -> T.EventM Name (T.Next AppState)
-appEvent s ev = case ev of
+appEvent s (T.VtyEvent ev) =
+  case view asAppMode s of
+    BrowseMail -> handle (Proxy :: Proxy 'BrowseMail) s ev
```

Comment:
Is there a way I can simplify this? I first thought I could write a function which can go from value of type `Mode` to a value of kind `Mode`(???)